README.md
# battery-monitoring-v0_4
Program Battery Monitoring (Node.js - Express - React). <br />
This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).

## Setup Software
1. Clone (atau download zip)
2. Dengan terminal Powershell #1 yang terdapat di dalam VSCode, pada direktori root (yang ada README.md nya - seharusnya tidak perlu diubah bila folder ini yang dibuka di VSCode), ketik ```npm install```
3. Dengan terminal Powershell #2 (split atau new terminal), pergi ke direktori backend dengan ```cd backend```, lalu ```node node_server.js```
4. Kembali ke root pada terminal #1 dengan ```cd ../```, lalu ```npm start```
5. Cek di browser, http://localhost:3000 untuk UI React, http://localhost:5000/buffer untuk buffer data nya.<br />
    Sementara ini, program dibuat untuk single user. Login dengan username "root" dan password "password".

## Setup Hardware
1. Pastikan Arduino dengan Battery Monitoring pada kondisi default ada di COM4, dengan baudRate 9600. Jika bukan di COM4, ganti bagian ini di file node_server.js:
    ```
    //deklarasi objek serial
    var portName = "COM4";
    var myPort = new serialport(portName,{
        baudRate:9600
    });
    ```
2. Sejauh ini cuma itu sih

## Menu yang Tersedia
1. Home <br /> Sebagai halaman Welcome umum.
2. Battery Monitoring <br /> Sebagai calon tempat menampilkan grafik per data
3. Battery Monitoring (Existing) <br /> Sistem Existing dari konsep awal (v0-1)
4. Read CSV <br /> Untuk membaca file CSV lokal. Harus menampung Header, dan saat ini baru dicoba untuk Timestamp dan 4 buah Voltage
5. Eng. Unit Setup <br /> Sebagai fitur untuk mengatur Scaling tampilan grafik. Untuk sementara, Versi 0.4 ini memungkinkan scaling untuk data dummy saja.
6. Logout <br /> Jelas untuk logout

## Link Router yang Digunakan
1. [Battery Monitoring](http://localhost:3000), sebagai aplikasi UI utama.
2. [Server Dummy](http://localhost:5000/api/buffer), sebagai tempat sementara data yang dikirim dari level 0

## Changelog
Lihat CHANGELOG.md.

## Useful Sources: 
Silakan cek file DAFPUS.md untuk mengetahui sumber di balik trik-trik pemrograman yang telah dipelajari.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.<br/>
**FORMAT TANGGAL : YYYY-MM-DD <br/>**

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

### Reminder
- Cek urutan variabel di halaman ReadCSV. Saat ini, timestamp ditempatkan pada kolom terakhir di dalam variabel dataset lokal milik halaman tersebut.
- Pikirkan cara untuk melakukan penskalaan pada variabel selain dummy. Fungsi ScaleDummy() mudah diimplementasikan karena datanya ada dalam satu chart, dan hanya chart itu saja. Dataset lain terpencar dalam tiga device yang berbeda, sehingga fungsi scaling nya perlu dibuat adaptif terhadap pilihan chart aktif.

## [0.4] - 2020-09-22

### Added
- Fitur Scaling pada bagian Eng. Unit Setup sekarang sudah difungsikan, tapi saat ini hanya untuk data dummy saja
- Kemampuan untuk menerima data lebih dari satu baterai. Versi ini dicobakan untuk data dari 2 device ADC yang terhubung dengan AI Signal Injector.
- Melengkapi komentar-komentar pada program (masih dalam proses, belum selesai)

### Changed
- Nama device disesuaikan dengan kasus nyata, tetapi dengan sinyal uji dari Signal Injector : Battery (4 tegangan + 1 arus), Motor_Stepper (2 tegangan), dan Controller (2 tegangan). Device Dummy masih ada.
- Jumlah data yang ditampilkan untuk Device pada chart disesuaikan

### Removed
- Data statis pada device selain device pertama dibuang (karena sudah ditambahkan data dari Signal Injector)

## [0.3] - 2020-09-15

### Added
- Folder ./backend/standalones/csvLogs yang memuat contoh-contoh file CSV yang bisa dibaca
- Program mandiri node_csv sebagai pembelajaran pembacaan CSV di node.js
- Halaman baru : Read CSV (sudah dengan isinya) dan Eng. Unit Setup (belum diisi)
- Melengkapi komentar-komentar pada program (masih dalam proses, belum selesai)

### Changed
- Referensi-referensi komponen DOM HTML diganti cara aksesnya menjadi ala React. <br /> 
    Contohnya ```document.getElementById``` menjadi ```this.<Sebuah nama Ref>.current```.
- Komponen Selector sekarang melewatkan prop berupa id device terpilih ke dalam ChartComp, dan interaksi update ChartComp didasarkan pada prop ini.

## [0.2] - 2020-09-09

### Added
- Halaman login, dan layout dasar dengan konsep React-Router
- User pertama "root" sebagai admin aplikasi
- Folder ./backend/standalones yang memuat snippet-snippet program yang berdiri sendiri. Berlaku juga sebagai folder sandboxing
- Folder  ./src/css yang memuat css global untuk seluruh komponen/layout
- Penggunaan (React-)Bootstrap untuk elemen-elemen UI

### Changed
- Data dummy menjadi data serial (Arduino di COM4)
- Komponen chart sekarang berhenti melakukan polling data dari server jika halaman yang memuatnya tidak aktif (di unmount)
- File aplikasi utama (App.js) sekarang tidak memuat MainWindow (komponen berisi chart), tetapi Router - yaitu modul yang menampilkan konten tergantung url aktif, yang dapat diubah dengan klik menu di bawah judul.

## [0.1] - 2020-09-04

### Added
- Fitur pembaca data serial (generator angka random di comment)
- [Sebuah server dummy](http://localhost:5000) yang hanya menampung GET data dari serial
- Folder ./backend untuk memuat aplikasi yang harus berjalan di balik layar (server)

### Changed
- Data dummy menjadi data serial (Arduino di COM4)
- Perlu 2 terminal untuk menjalankan dua file yang berbeda (node untuk server, npm start untuk client)

## [0.0] - 2020-09-02

### Added
- Upload awal

### Changed
- Perubahan basis program dari tampilan web statis (HTML-JS-PHP) ke React-app

### Removed
- Halaman login pada web statis


backend/node_server.js
/************************** Import library/fungsi ****************************/
//library serialport
var serialport = require('serialport');

//library untuk server tempat buffer serta koneksi mysql
const express = require('express');

//library untuk (format) timestamp
var moment  = require('moment');

/************************ Deklarasi objek/variabel ***************************/
//Buffer data battery monitoring dari serial
var Buff_Unit_A1 = [
    [0],
    [0],
    [0],
    [0],
    [0],
    [0],
    [0],
    [0],
    [0]
];
var timestampBuff_Unit_A1 = [
    [0],
    [0],
    [0],
    [0],
    [0],
    [0],
    [0],
    [0],
    [0]
];
var currentBuff_Unit_A1 = [[0],[0],[0],[0],[0],[0],[0],[0],[0]];
const dataLimit = 20;
const normalDataLimit = 100;
// Jumlah data, atau jumlah kolom
// ada 4 tegangan 1 arus (baterai) + 2 (motor) + 2 (controller)
// 5 data + 2 data + 2 data = 9
const seriesCount = 9;
var iter = 0;

//deklarasi objek serial
var portName = "COM6";
var myPort = new serialport(portName,{
    baudRate:9600
});

//deklarasi objek server
const app = express();
const port = 5000;

/************************ Deklarasi fungsi/event ***************************/
function StoreToBuffer(currentBuff,Buff,Timestamp){
    for(var i=0;i<seriesCount;i++){
        if((Buff[i].length)>dataLimit){
            Buff[i].shift();
            Buff[i][dataLimit] = currentBuff[i][0];
            Timestamp[i].shift();
            Timestamp[i][dataLimit] = String(moment().format('hh:mm:ss'));
        }
        else{
            Buff[i][Buff[i].length] = currentBuff[i][0];
            Timestamp[i][Timestamp[i].length] = String(moment().format('hh:mm:ss'));
        }
    }
    return [Buff,Timestamp];
}

//event listener untuk mengoper data dari serial ke server
function expressGETBuffer(){
    app.get('/api/buffer',(req,res) =>{
        var devices = [
            {
                id:1, name: 'Battery', 
                voltage1: Buff_Unit_A1[0],
                voltage2: Buff_Unit_A1[1],
                voltage3: Buff_Unit_A1[2],
                voltage4: Buff_Unit_A1[3],
                arus: Buff_Unit_A1[4],
                timestamp: timestampBuff_Unit_A1[0]
            },
            {
                id:2, name: 'Motor_Stepper',
                voltage1: Buff_Unit_A1[5],
                voltage2: Buff_Unit_A1[6],
                voltage3: [],
                voltage4: [],
                arus: [],
                timestamp: timestampBuff_Unit_A1[0]
            },
            {
                id:3, name: 'Controller', 
                voltage1: Buff_Unit_A1[7],
                voltage2: Buff_Unit_A1[8],
                voltage3: [],
                voltage4: [],
                arus: [],
                timestamp: timestampBuff_Unit_A1[0]
            },
        ];
        res.json(devices);
    });
    //console.log("2a. GET to Server");
}

//event Listener bila ada serial masuk dari Arduini
myPort.on("data", (line) => {
    //console.log("1a. Serial data Acquired");
    iter +=1;
    //ubah tipe data Buffer Javascript jadi Array
    bufferArray = [...line];

    //satukan data buffer menjadi integer
    V1_int = (bufferArray[1]<<8) | bufferArray[0];
    V2_int = (bufferArray[3]<<8) | bufferArray[2];
    V3_int = (bufferArray[5]<<8) | bufferArray[4];
    V4_int = (bufferArray[7]<<8) | bufferArray[6];
    Arus_int = (bufferArray[9]<<8) | bufferArray[8];
    VM1_int = (bufferArray[11]<<8) | bufferArray[10];
    VM2_int = (bufferArray[13]<<8) | bufferArray[12];
    VC1_int = (bufferArray[15]<<8) | bufferArray[14];
    VC2_int = (bufferArray[17]<<8) | bufferArray[16];

    //konversi integer ke float    
    V1 = V1_int/100.0; if(V1>normalDataLimit) V1=0;
    V2 = V2_int/100.0; if(V2>normalDataLimit) V2=0;
    V3 = V3_int/100.0; if(V3>normalDataLimit) V3=0;
    V4 = V4_int/100.0; if(V4>normalDataLimit) V4=0;
    
    Arus = Arus_int/100.0;

    VM1 = VM1_int/100.0; if(VM1>normalDataLimit) VM1=0;
    VM2 = VM2_int/100.0; if(VM2>normalDataLimit) VM2=0;
    VC1 = VC1_int/100.0; if(VC1>normalDataLimit) VC1=0;
    VC2 = VC2_int/100.0; if(VC2>normalDataLimit) VC2=0;
    //console.log("1b. Serial data Converted");

    //masukkan ke buffer agar dapat diakses chart dan server
    currentBuff_Unit_A1 = [[V1],[V2],[V3],[V4],[Arus],[VM1],[VM2],[VC1],[VC2]];
    StoreToBuffer(currentBuff_Unit_A1,Buff_Unit_A1,timestampBuff_Unit_A1);
    //console.log("1c. Serial data Stored to Buffer");
    
    console.log(bufferArray);
    //debug
    /*
    console.log(bufferArray);
    console.log(`iter:${iter}`);
    console.log(`Arus : ${Arus}`);
    console.log('V1 : '+ String(V1));
    console.log(`V2 : ${V2}`);
    */
});

myPort.on("close",() => {
    console.log("RECON coz close");
    myPort.resume();
});

myPort.on("error",() => {
    console.log("RECON coz error");
    myPort.resume();
});
/****************************** Main Loop ***********************************/
//untuk polling data tiap x detik
function pollData(){
    console.log(`\n-------------------- NEW LOOP #${iter} --------------------\n`);
    pollCharacter = [0x11];
    //console.log("1. Polled Serial Port");
    myPort.write(pollCharacter);
    //console.log("2. GET Procedure");
    expressGETBuffer();
}
setInterval(pollData,1000);

/******************** Finalisasi (nyalakan server) *************************/
app.listen(port, () => console.log(`Server started on port ${port}`));